core/src/main/java/com/graphhopper/routing/Router.java

```diff
@@ -102,111 +102,210 @@ public Router(GraphHopperStorage ghStorage, LocationIndex locationIndex,
+                ViaRouting.Result result = ViaRouting.calcPaths(request.getPoints(), queryGraph, qResults, weighting.getFlagEncoder().getAccessEnc(), pathCalculator, request.getCurbsides(), forceCurbsides, request.getHeadings(), passThrough);
+
+                if (request.getPoints().size() != result.paths.size() + 1)
+                    throw new RuntimeException("There should be exactly one more point than paths. points:" + request.getPoints().size() + ", paths:" + result.paths.size());
+
+                // here each path represents one leg of the via-route and we merge them all together into one response path
+                ResponsePath responsePath = concatenatePaths(request, weighting, queryGraph, result.paths, getWaypoints(qResults));
+                responsePath.addDebugInfo(result.debug);
+                ghRsp.add(responsePath);
+                ghRsp.getHints().putObject("visited_nodes.sum", result.visitedNodes);
+                ghRsp.getHints().putObject("visited_nodes.average", (float) result.visitedNodes / (qResults.size() - 1));
+                return result.paths;
+            }
+        } catch (MultiException ex) {
+            for (Throwable t : ex.getErrors()) {
+                ghRsp.addError(t);
+            }
+            return emptyList();
         } catch (IllegalArgumentException ex) {
             ghRsp.addError(ex);
-            return Collections.emptyList();
+            return emptyList();
         }
     }
 
-    public RoutingAlgorithmFactory getAlgorithmFactory(String profile, boolean disableCH, boolean disableLM) {
-        if (chEnabled && disableCH && !routerConfig.isCHDisablingAllowed()) {
-            throw new IllegalArgumentException("Disabling CH is not allowed on the server side");
+    private Weighting createWeighting(Profile profile, PMap requestHints, List<GHPoint> points, boolean forCH) {
+        if (forCH) {
+            // todo: do not allow things like short_fastest.distance_factor or u_turn_costs unless CH is disabled
+            // and only under certain conditions for LM
+
+            // the request hints are ignored for CH as we cannot change the profile after the preparation like this.
+            // the weighting here has to be created the same way as we did when we created the weighting for the preparation
+            return weightingFactory.createWeighting(profile, new PMap(), false);
+        } else {
+            Weighting weighting = weightingFactory.createWeighting(profile, requestHints, false);
+            if (requestHints.has(Parameters.Routing.BLOCK_AREA)) {
+                FlagEncoder encoder = encodingManager.getEncoder(profile.getVehicle());
+                GraphEdgeIdFinder.BlockArea blockArea = GraphEdgeIdFinder.createBlockArea(ghStorage, locationIndex,
+                        points, requestHints, DefaultEdgeFilter.allEdges(encoder));
+                weighting = new BlockAreaWeighting(weighting, blockArea);
```

Comment:
Can we somehow use this Router class (or something else in the future) to avoid the BlockAreaWeighting creation in SPTResource and IsochroneResource?

Reply:
To me it looks like the block area stuff seems to belong to the weighting factory (but of course its not entirely clear because we need the 'points' here...). If all this was in weighting factory SPT/Isochrone could just use it.

`Router` is kind of 'the thing that does routing when everything is ready', so in theory it might be the right place to do map matching, isochrone etc. as well. Or maybe not and we will just have similar classes next to `Router` (`MapMatcher`, `Isochrone*thing*` etc.). They all have in common that they need to have access to the same 'infrastructure' like the routing graphs, profiles etc.